### PR TITLE
Promote Pod+PodStatus resource lifecycle test - +4 endpoint coverage

### LINKED ISSUE #90936: Write CoreV1Pod and CoreV1PodStatus lifecycle test - +4 coverage
# Progress `[7/8]`
* [x] Original approval issue: #88545
* [x]  Test approval issue: [kubernetes/kubernetes#90936](https://github.com/kubernetes/kubernetes/issues/90936)
* [x]  [APISnoop PR](https://github.com/cncf/apisnoop/pull/290)
* [x]  Test pr: [PR](https://github.com/kubernetes/kubernetes/pull/90942) - Reverted due to flakes
* [x] New Test [PR](https://github.com/kubernetes/kubernetes/pull/93459)
* [x]  Two weeks soak start date: 7 November 2020 [testgrid-link](https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&width=5&include-filter-by-regex=should%20run%20through%20the%20lifecycle%20of%20Pods%20and%20PodStatus) 
* [ ]  Two weeks soak end date: 20 November 2020
* [x] Test promotion pr: [#96485](https://github.com/kubernetes/kubernetes/pull/96485)

This issue is created to allow edit by @Riaankl

/wip
/hold

# Identify an untested feature Using APISnoop
According to this APIsnoop query, there are still some remaining Pod endpoints which are untested.

```
SELECT
  operation_id,
  -- k8s_action,
  -- path,
  -- description,
  kind
  FROM untested_stable_core_endpoints
  where path not like '%volume%'
  and kind like 'Pod'
  -- and kind like ''
  -- and operation_id ilike '%%'
 ORDER BY kind,operation_id desc
 -- LIMIT 25
       ;
```

```
            operation_id             | kind 
-------------------------------------|------
 replaceCoreV1NamespacedPodStatus    | Pod
 readCoreV1NamespacedPodStatus       | Pod
 listCoreV1PodForAllNamespaces       | Pod
 deleteCoreV1CollectionNamespacedPod | Pod
(4 rows)
```

# Use API Reference to Lightly Document the Feature
* [Kubernetes API Reference Docs](https://kubernetes.io/docs/reference/kubernetes-api/)
* [client-go - Pod](https://github.com/kubernetes/client-go/blob/master/kubernetes/typed/core/v1/pod.go)

# The mock test
## Test outline
1. Create a Pod with a static label
2. Patch the Pod with a new Label and updated data
3. Get the Pod to ensure it's patched
4. Read the Pod's status
5. Replace the Pod's status Ready condition to False
6. Read the Pod's status to check if Ready is False
7. List all Pods in all Namespaces find the Pod(1) ensure that the Pod is found and is patched
8. Delete Namespaced Pod(1) via a Collection with a LabelSelector
9. Get the Pod to check that it's deleted

## Example in Go
```go
package main
import (
  "encoding/json"
  "time"
  "fmt"
  "flag"
  "os"
  v1 "k8s.io/api/core/v1"
  "k8s.io/client-go/dynamic"
  "k8s.io/apimachinery/pkg/runtime/schema"
  metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
  "k8s.io/client-go/kubernetes"
  "k8s.io/apimachinery/pkg/types"
  "k8s.io/client-go/tools/clientcmd"
)
func main() {
  // uses the current context in kubeconfig
  kubeconfig := flag.String("kubeconfig", fmt.Sprintf("%v/%v/%v", os.Getenv("HOME"), ".kube", "config"), "(optional) absolute path to the kubeconfig file")
  flag.Parse()
  config, err := clientcmd.BuildConfigFromFlags("", *kubeconfig)
  if err != nil {
      fmt.Println(err)
      return
  }
  // make our work easier to find in the audit_event queries
  config.UserAgent = "live-test-writing"
  // creates the clientset
  ClientSet, _ := kubernetes.NewForConfig(config)
  DynamicClientSet, _ := dynamic.NewForConfig(config)
  podResource := schema.GroupVersionResource{Group: "", Version: "v1", Resource: "pods"}
  testNs := "default"
  testPodName := "pod-test"
  testPodImage := "nginx"
  testPodImage2 := "nginx"
  testPod := v1.Pod{
      ObjectMeta: metav1.ObjectMeta{
          Name: testPodName,
          Labels: map[string]string{
              "test-pod-static": "true",
          },
      },
      Spec: v1.PodSpec{
          Containers: []v1.Container{
              {
                 Name: testPodName,
                 Image: testPodImage,
              },
          },
      },
  }
  // create Pod with a static label
  _, err = ClientSet.CoreV1().Pods(testNs).Create(&testPod)
  if err != nil {
      fmt.Println(err)
      return
  }
  fmt.Println("[status] created Pod")
  // setup a watch for the RC
  podWatch, err := ClientSet.CoreV1().Pods(testNs).Watch(metav1.ListOptions{LabelSelector: "test-pod-static=true"})
  if err != nil {
      fmt.Println(err)
      return
  }
  podWatchChan := podWatch.ResultChan()
  fmt.Println("[status] watching for Pod to be ready")
  for event := range podWatchChan {
      podEvent, ok := event.Object.(*v1.Pod)
      if ok != true {
         fmt.Println("Unable to fix type")
         return
      }
      if podEvent.Status.Phase == "Running" {
         break
      }
  }
  fmt.Println("[status] Pod is Ready")
  // patch the Pod with a new Label and updated data
  podPatch, err := json.Marshal(map[string]interface{}{
      "metadata": map[string]interface{}{
          "labels": map[string]string{
              "podtemplate": "patched",
          },
      },
      "spec": map[string]interface{}{
          "containers": []map[string]interface{}{
              {
                  "name": testPodName,
                  "image": testPodImage2,
              },
          },
      },
  })
  if err != nil {
      fmt.Println(err)
      return
  }
  _, err = ClientSet.CoreV1().Pods(testNs).Patch(testPodName, types.StrategicMergePatchType, []byte(podPatch))
  if err != nil {
      fmt.Println(err)
      return
  }
  // get the Pod and ensure it's patched
  pod, err := ClientSet.CoreV1().Pods(testNs).Get(testPodName, metav1.GetOptions{})
  if err != nil {
      fmt.Println(err)
      return
  }
  if pod.ObjectMeta.Labels["test-pod-static"] != "true" || pod.Spec.Containers[0].Image != testPodImage2 {
      fmt.Println("[error] patching of Pod failed")
      return
  }
  // get pod status
  podStatusUnstructured, err := DynamicClientSet.Resource(podResource).Namespace(testNs).Get(testPodName, metav1.GetOptions{}, "status")
  if err != nil {
      fmt.Println(err)
      return
  }
  podStatusUjson, _ := json.Marshal(podStatusUnstructured)
  var podStatus v1.Pod
  json.Unmarshal(podStatusUjson, &podStatus)
  // replace the Pod's status Ready condition to False
  podStatusUpdated := podStatus
  podStatusFieldPatchCount := 0
  podStatusFieldPatchCountTotal := 2
  for pos, cond := range podStatusUpdated.Status.Conditions {
      if (cond.Type == "Ready" && cond.Status == "True") || (cond.Type == "ContainersReady" && cond.Status == "True") {
          podStatusUpdated.Status.Conditions[pos].Status = "False"
          podStatusFieldPatchCount++
      }
  }
  if podStatusFieldPatchCount != podStatusFieldPatchCountTotal {
      fmt.Println("[error] failed to patch all relevant Pod conditions")
      return
  }
  _, err = ClientSet.CoreV1().Pods(testNs).UpdateStatus(&podStatusUpdated)
  if err != nil {
      fmt.Println(err)
      return
  }
  fmt.Println("[status] updated PodStatus")
  // list all Pods and get their status to ensure it's Ready condition is False
  podsList, err := ClientSet.CoreV1().Pods("").List(metav1.ListOptions{LabelSelector: "test-pod-static=true"})
  if err != nil {
      fmt.Println(err)
      return
  }
  fmt.Println("[status] fetched all Pods by LabelSelector")
  podStatusFieldPatchCount = 0
  podStatusFieldPatchCountTotal = 2
  for _, podItem := range podsList.Items {
      for _, cond := range podItem.Status.Conditions {
          if (cond.Type == "Ready" && cond.Status == "False") || (cond.Type == "ContainersReady" && cond.Status == "False") {
              podStatusFieldPatchCount++
          }
      }
  }
  if podStatusFieldPatchCount != podStatusFieldPatchCountTotal {
      fmt.Printf("[error] failed to update PodStatus - %v/%v conditions failed to update (%v, %v)", podStatusFieldPatchCount, podStatusFieldPatchCountTotal, "Ready", "ContainersReady")
      return
  }
  fmt.Println("[status] PodStatus was updated successful")
  // delete the Pod via a Collection with a LabelSelector
  err = ClientSet.CoreV1().Pods(testNs).DeleteCollection(&metav1.DeleteOptions{}, metav1.ListOptions{LabelSelector: "test-pod-static=true"})
  if err != nil {
      fmt.Println(err)
      return
  }
  fmt.Println("[status] watching for Pod to be not Ready")
  podEventChannel:
  for event := range podWatchChan {
      podEvent, ok := event.Object.(*v1.Pod)
      if ok != true {
          fmt.Println("Unable to fix type")
          return
      }
      podStatusFieldPatchCount := 0
      podStatusFieldPatchCountTotal := 2
      for _, cond := range podEvent.Status.Conditions {
          if (cond.Type == "Ready" && cond.Status == "False") || (cond.Type == "ContainersReady" && cond.Status == "False") {
              podStatusFieldPatchCount++
          }
      }
      if podStatusFieldPatchCount == podStatusFieldPatchCountTotal {
          break podEventChannel
      }
  }
  time.Sleep(5 * time.Second)
  fmt.Println("[status] Pod no longer available")
  // fetch the Pod to check if it's deleted
  _, err = ClientSet.CoreV1().Pods(testNs).Get(testPodName, metav1.GetOptions{})
  if err == nil {
      fmt.Println("[error] Pod still available after deletion; failed to delete Pod")
      return
  }
  // write test here
  fmt.Println("[status] complete")
}
```

```go
[status] created Pod
[status] watching for Pod to be ready
[status] Pod is Ready
[status] updated PodStatus
[status] fetched all Pods by LabelSelector
[status] PodStatus was updated successful
[status] watching for Pod to be not Ready
[status] Pod no longer available
[status] complete
```

# Verify Increase it Coverage with APISnoop
Discover useragents:

```
select distinct useragent from audit_event where bucket='apisnoop' and useragent not like 'kube%' and useragent not like 'coredns%' and useragent not like 'kindnetd%' and useragent like 'live%';
```

```
     useragent     
-------------------
 live-test-writing
(1 row)
```

List endpoints hit by the test:

```
select * from endpoints_hit_by_new_test where useragent like 'live%'; 
```

```
     useragent     |            operation_id             | hit_by_ete | hit_by_new_test 
-------------------|-------------------------------------|------------|-----------------
 live-test-writing | readCoreV1NamespacedPodStatus       | f          |               1
 live-test-writing | readCoreV1NamespacedPod             | t          |               2
 live-test-writing | deleteCoreV1CollectionNamespacedPod | f          |               2
 live-test-writing | patchCoreV1NamespacedPod            | t          |               2
 live-test-writing | listCoreV1PodForAllNamespaces       | f          |               1
 live-test-writing | createCoreV1NamespacedPod           | t          |               2
 live-test-writing | replaceCoreV1NamespacedPodStatus    | f          |               2
 live-test-writing | listCoreV1NamespacedPod             | t          |               1
(8 rows)
```

Display endpoint coverage change:

```
select * from projected_change_in_coverage;
```

```
   category    | total_endpoints | old_coverage | new_coverage | change_in_number 
---------------|-----------------|--------------|--------------|------------------
 test_coverage |             438 |          190 |          194 |                4
(1 row)
```

# Final notes
If a test with these calls gets merged, ****test coverage will go up by 4 points****

This test is also created with the goal of conformance promotion.

/sig testing
/sig architecture
/area conformance

### PR DESCRIPTION
**Pull request by riaankleinhans**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it:**
This PR adds a test to test the following untested endpoints:
* replaceCoreV1NamespacedPodStatus
* readCoreV1NamespacedPodStatus
* listCoreV1PodForAllNamespaces
* deleteCoreV1CollectionNamespacedPod

**Which issue(s) this PR fixes:**
Fixes #90936
[Testgrid link](https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&width=5&include-filter-by-regex=should%20run%20through%20the%20lifecycle%20of%20Pods%20and%20PodStatus&graph-metrics=test-duration-minutes)

**Special notes for your reviewer:**
Adds +4 endpoint test coverage (good for conformance)

**Does this PR introduce a user-facing change?:**
```
NONE

```

**Release note:**
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:**
```
NONE

```

/sig testing
/sig architecture
/area conformance

### CODE REVIEW

**Comment by hh** [comment]
/test 
pull-kubernetes-node-e2e

**Comment by oomichi** [comment]
We need to wait 9 days more at least to check the test stability..

**Review by hasheddan — COMMENTED** [review]
/milestone v1.20

**Comment by hh** [comment]
@oomichi We usually prep these in advance so that on the day we reach enough non-flakes. The PR is ready to merge. :)

**Comment by Riaankl** [comment]
A prefect run in T[estgrid](https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&width=5&include-filter-by-regex=should%20run%20through%20the%20lifecycle%20of%20Pods%20and%20PodStatus)
Can we please get a /LGTM & Approve
/cc @smarterclayton @oomichi

**Comment by oomichi** [comment]
This test is failed in a single job today as https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce/1329047804261175296

```
Nov 18 13:29:04.265: FAIL: failed to see Pod pod-test in namespace pods-5897 running
Unexpected error:
    <*errors.errorString | 0xc0002781f0>: {
        s: "timed out waiting for the condition",
    }
    timed out waiting for the condition
occurred

Full Stack Trace
k8s.io/kubernetes/test/e2e/common.glob..func18.13()
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/common/pods.go:941 +0xbd7
k8s.io/kubernetes/test/e2e.RunE2ETests(0xc003212a80)
	_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/e2e.go:130 +0x36c
k8s.io/kubernetes/test/e2e.TestE2E(0xc003212a80)
	_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/e2e_test.go:144 +0x2b
testing.tRunner(0xc003212a80, 0x4fa0500)
	/usr/local/go/src/testing/testing.go:1127 +0xef
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1178 +0x386
```

**Comment by hh** [comment]
We had a single flake in a couple weeks.
It seems we were able to reach PodScheduled, but just needed a bit more time for it to reach state running.
Fix: [Increase timeout for pod lifecycle test to reach pod status=ready #96691](https://github.com/kubernetes/kubernetes/pull/96691)

```
[It] should run through the lifecycle of Pods and PodStatus
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/common/pods.go:887
ï¿½[1mSTEPï¿½[0m: creating a Pod with a static label
ï¿½[1mSTEPï¿½[0m: watching for Pod to be ready
Nov 18 13:28:04.312: INFO: observed Pod pod-test in namespace pods-5897 in phase Pending conditions []
Nov 18 13:28:04.360: INFO: observed Pod pod-test in namespace pods-5897 in phase Pending conditions [{PodScheduled True 0001-01-01 00:00:00 +0000 UTC 2020-11-18 13:28:04 +0000 UTC  }]
Nov 18 13:29:04.265: FAIL: failed to see Pod pod-test in namespace pods-5897 running
Unexpected error:
    <*errors.errorString | 0xc0002781f0>: {
        s: "timed out waiting for the condition",
    }
    timed out waiting for the condition
occurred
```

**Comment by hh** [comment]
In my opinion, the reason for this flake is clear, we needed more time.
We've addressed that and I still put forward that we should approve this for 1.20.

**Comment by oomichi** [comment]
```
Nov 18 13:28:04.312: INFO: observed Pod pod-test in namespace pods-5897 in phase Pending conditions []
Nov 18 13:28:04.360: INFO: observed Pod pod-test in namespace pods-5897 in phase Pending conditions [{PodScheduled True 0001-01-01 00:00:00 +0000 UTC 2020-11-18 13:28:04 +0000 UTC  }]
Nov 18 13:29:04.265: FAIL: failed to see Pod pod-test in namespace pods-5897 running
```
After detecting PodScheduled at 13:28:04, the test didn't show the pod status at all before the test was failed 1 minute later at line 941.
That means `event.Object.(*v1.Pod)` didn't return true of ok value:
```
 928                 _, err = watchtools.Until(ctx, podsList.ResourceVersion, w, func(event watch.Event) (bool, error) {
 929                         if pod, ok := event.Object.(*v1.Pod); ok {
 930                                 found := pod.ObjectMeta.Name == testPod.ObjectMeta.Name &&
 931                                         pod.ObjectMeta.Namespace == testNamespaceName &&
 932                                         pod.Labels["test-pod-static"] == "true" &&
 933                                         pod.Status.Phase == v1.PodRunning
 934                                 if !found {
 935                                         framework.Logf("observed Pod %v in namespace %v in phase %v conditions %v", pod.ObjectMeta.Name, pod.ObjectMeta.Namespace, pod.Status.Phase, pod     .Status.Conditions)
 936                                 }
 937                                 return found, nil
 938                         }
 939                         return false, nil
 940                 })
 941                 framework.ExpectNoError(err, "failed to see Pod %v in namespace %v running", testPod.ObjectMeta.Name, testNamespaceName)
```
Can we get the reason why the value `ok` becomes false?

**Comment by BobyMCbobs** [comment]
> Can we get the reason why the value `ok` becomes false?

@oomichi, without checking of the type-assertion (`ok` parts): I believe that it would have a panic because the sub-properties (i.e `pod.Status.Phase`) would show as invalid.

**Comment by oomichi** [comment]
>>    Can we get the reason why the value ok becomes false?
>
> @oomichi, without checking of the type-assertion (ok parts): I believe that it would have a panic because the sub-properties (i.e pod.Status.Phase) would show as invalid.

Then I guess it is difficult to say `this flake is clear, we needed more time.` because we didn't see any concrete process when the timeout happened.

**Comment by heyste** [comment]
@oomichi,  Checking the [graph metrics for this e2e test](https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&width=5&include-filter-by-regex=should%20run%20through%20the%20lifecycle%20of%20Pods%20and%20PodStatus&graph-metrics=test-duration-minutes) the test run that flaked looks to be due to fluctuations in the cluster environment. Three other tests also failed in the same run.

There are at least 6 test runs where the time taken has been at least 2 to 3 times as long as the running average from what I see from the graph. I believe the current timeout is just too optimistic for when the cluster environment is not running as smoothly as it has been for most of the test runs.

**Comment by hh** [comment]
The other tests that fail in the same job, are all showing a similar spike in time. I suspect the cluster load was just higher than usual, and needed slightly more time.

image

**Review by spiffxp — APPROVED** [review]
/approve
/lgtm

**Comment by spiffxp** [comment]
I'm willing to write off the failure as a fluke since there is only one occurrence.

I do think the test's logging leaves a little to be desired in troubleshooting which specific if statement or which boolean check evaluated to false:
https://github.com/kubernetes/kubernetes/blob/540e41c825f71d5ac67bff97902c8e689784a474/test/e2e/common/pods.go#L929-L939
- was the watch event not for a Pod?
- was the Pod missing the `test-pod-static: true` label?
- was the Pod's Phase not PodRunning?

**Comment by spiffxp** [comment]
/test pull-kubernetes-e2e-gce-100-performance
bazel build flake while pulling images
```W1123 18:06:32.387] 2020/11/23 18:06:32 Image pull was unsuccessful: unable to save remote image k8s.gcr.io/build-image/debian-iptables@sha256:e30919918299988b318f0208e7fd264dee21a6be9d74bbd9f7fc15e78eade9b4: unable to write image layers: unable to write image layer: unable to write the contents of layer 1 to /bazel-scratch/.cache/bazel/_bazel_prow/48d5366022b4e3197674c8d6e2bee219/external/debian-iptables-amd64/image/001.tar.gz: read tcp 10.34.14.3:34084->172.217.214.128:443: read: connection reset by peer```

**Comment by heyste** [comment]
/test pull-kubernetes-bazel-test
flake
```
FAIL: //staging/src/k8s.io/apiserver/pkg/server:go_default_test (run 2 of 2) (see /bazel-scratch/.cache/bazel/_bazel_root/7989b31489f31aee54f32688da2f0120/execroot/io_k8s_kubernetes/bazel-out/k8-fastbuild/testlogs/staging/src/k8s.io/apiserver/pkg/server/go_default_test/run_2_of_2/test.log)
INFO: From Testing //staging/src/k8s.io/apiserver/pkg/server:go_default_test (run 2 of 2):
==================== Test output for //staging/src/k8s.io/apiserver/pkg/server:go_default_test (run 2 of 2):
W1123 18:29:10.351356  181906 authorization.go:47] Authorization is disabled
W1123 18:29:10.351750  181906 authentication.go:40] Authentication is disabled
W1123 18:29:10.361761  181906 authorization.go:47] Authorization is disabled
	staging/src/k8s.io/apiserver/pkg/server/config_test.go:110 +0x7e9
testing.tRunner(0xc0000cf080, 0x2b74f38)
	GOROOT/src/testing/testing.go:1127 +0x203
created by testing.(*T).Run
	GOROOT/src/testing/testing.go:1178 +0x797
--- FAIL: TestNewWithDelegate (0.10s)
    --- FAIL: TestNewWithDelegate/http://127.0.0.1:44977/healthz (0.00s)
        config_test.go:210: HTTP/1.1 500 Internal Server Error
            Content-Length: 334
```

**Comment by mmerkes** [comment]
Looks like this test is causing the NodeConformance tests to fail. Setup for the NodeConformance is different, so we'll need to figure that out.

https://github.com/kubernetes/kubernetes/issues/97130

**Pull request merged**
